AdvancedDLSupport/ImplementationGenerators/Wrappers/SpanMarshallingWrapper.cs

```diff
@@ -1,135 +1,240 @@
+
+                    var getPinnableReferenceMethod = paramType.GetMethod(nameof(Span<byte>.GetPinnableReference), BindingFlags.Public | BindingFlags.Instance);
+
+                    il.EmitLoadArgumentAddress(i);
+                    il.EmitCallDirect(getPinnableReferenceMethod);
+                    il.EmitDuplicate();
+                    il.EmitSetLocalVariable(pinnedLocal);
+                    il.EmitConvertToNativeInt();
+                }
+                else
+                {
+                    il.EmitLoadArgument(i);
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public override void EmitEpilogue(ILGenerator il, PipelineWorkUnit<IntrospectiveMethodInfo> workUnit)
+        {
+            var returnType = workUnit.Definition.ReturnType;
+
+            if (IsSpanType(returnType))
+            {
+                il.EmitConstantInt(GetNativeCollectionLengthMetadata(workUnit.Definition).Length);
+                il.EmitNewObject(returnType.GetConstructor(new[] { typeof(void*), typeof(int) }));
+            }
+        }
+
+        private NativeCollectionLengthAttribute GetNativeCollectionLengthMetadata(IntrospectiveMethodInfo info)
+        {
+            IReadOnlyList<CustomAttributeData> attributes = info.ReturnParameterCustomAttributes;
+
+            foreach (CustomAttributeData customAttributeData in attributes)
+            {
+                if (customAttributeData.AttributeType == typeof(NativeCollectionLengthAttribute))
+                {
+                    return customAttributeData.ToInstance<NativeCollectionLengthAttribute>();
+                }
+            }
+
+            throw new InvalidOperationException($"Method return type does not have required {nameof(NativeCollectionLengthAttribute)}");
+        }
+
+        /// <inheritdoc />
+        public override GeneratorComplexity Complexity => GeneratorComplexity.TransformsParameters | GeneratorComplexity.MemberDependent;
+
+        /// <inheritdoc />
+        public override bool IsApplicable(IntrospectiveMethodInfo member)
+        {
+            return IsSpanType(member.ReturnType) || member.ParameterTypes.Any(IsSpanType);
+        }
+
+        /// <summary>
+        /// Determines whether the <see cref="Type" /> provided is a generic span.
+        /// </summary>
+        /// <param name="type">The type to check.</param>
+        private static bool IsSpanType([NotNull] Type type)
+        {
+            if (type.IsGenericType)
+            {
+                var generic = type.GetGenericTypeDefinition();
+                return generic == typeof(Span<>) || generic == typeof(ReadOnlySpan<>);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines whether the <see cref="Type" /> provided contains references.
+        /// </summary>
+        /// <param name="type">The type to check.</param>
+        private static bool IsOrContainsReferences([NotNull] Type type)

```

Comment:
This function may need some additional checks to be completely valid, since it's more or less checking for compatibility with the `unmanaged` constraint. 

https://stackoverflow.com/questions/53968920/how-do-i-check-if-a-type-fits-the-unmanaged-constraint-in-c

Reply:
Splitting it out into an extension method might not be a bad idea either.

Reply:
Meh it's a private method so unless we use it elsewhere in the codebase (in which case I'll split it off into a separate file), we needn't worry.

Reply:
Meh it's a private method so unless we use it elsewhere in the codebase (in which case I'll split it off into a separate file), we needn't worry.

Reply:
Fixed.

Reply:
Adding a nested class for the test makes me want you to split it out even more :P

Reply:
Fixed.